### PR TITLE
Forward transactions to the expected leader instead of your own TPU port

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Using your own TPU port to forward transactions uses one of the hops, so if the current node is behind the tx will get dropped.

#### Summary of Changes

Get the expected leader by PoH time and forward to that leader.

Fixes #

### CODE REVIEW

File: core/src/send_transaction_service.rs [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+            .name("send-tx-svc".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+
+                if let Ok(transaction_info) = receiver.recv_timeout(Duration::from_secs(1)) {
+                    let address = leader_info
+                        .as_ref()
+                        .and_then(|recorder| recorder.1.lock().unwrap().leader_after_n_slots(0))
```
Comment (mvines) [inline core/src/send_transaction_service.rs:103]: Should we aim at the next slot instead?
```suggestion
                        .and_then(|recorder| recorder.1.lock().unwrap().leader_after_n_slots(1))
```

Comment (aeyakovenko) [reply core/src/send_transaction_service.rs:103]: Why?  Current leader should be the right slot unless they are done and they will send it to the next leader.

Comment (mvines) [reply core/src/send_transaction_service.rs:103]: Half the time you'll be in the last 50% of the current slot when this code runs, then include the network propagation time from RPC to the leader -- this takes us to very close to the end of the slot or into the next slot already by the time it reaches the target node.

If we miss the leader, yes they will send it to the next leader.
But if we hit the leader one slot early, they'll just hang on to the transaction.

EmpIrically I think this means that we're more likely to hit the actual leader and avoid the forward.

**Comment by CriesofCarrots** [comment]
Looks all right to me, pending the leader-slot decision. It's a shame about the duplicate `SendTransactionService`s, but I guess the alternative would be a new crate with just `PohRecorder` and `SendTransactionService` 😕

File: core/src/send_transaction_service.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+                match working_bank.get_signature_status_slot(signature) {
+                    None => {
+                        // Transaction is unknown to the working bank, it might have been
+                        // dropped or landed in another fork.  Re-send it
+                        info!("Retrying transaction: {}", signature);
+                        result.retried += 1;
+                        inc_new_counter_info!("send_transaction_service-retry", 1);
+                        Self::send_transaction(
+                            &send_socket,
+                            &tpu_address,
```
Comment (carllin) [inline core/src/send_transaction_service.rs]: I think this should also use Poh, updated here: https://github.com/solana-labs/solana/pull/12004/commits/fe265bfca81599cfd6b85497dc5413a1fad8b068

File: core/src/send_transaction_service.rs (hunk 1 above)
Comment (jstarry) [reply core/src/send_transaction_service.rs]: I think we should fanout the transaction to the current + next few (2-3) leaders

**Comment by sakridge** [comment]
Evidence of improvement?

File: core/src/send_transaction_service.rs (hunk 1 above)
Comment (mvines) [reply core/src/send_transaction_service.rs]: Ok.  I don't think we need to hold up this PR due to this thread.  The current patch is a definite improvement over the current situation, we can iterate more later if needed

**Comment by aeyakovenko** [comment]
Seems to be faster, but break and rpc nodes are seeing other issues. 

Not sure how to test this thoroughly.  @carllin can you try it with serum folks?

**Comment by carllin** [comment]
@aeyakovenko I've pinged them, no response yet. 

I ran some transactions from the Serum UI and pointed it toward the upgraded RPC node, and the response times seem reasonable, but the RPC nodes are not behind like they were on Wednesday, so may not be an accurate measure.

**Comment by aeyakovenko** [comment]
Want to land it?  This is doing the right thing, vs the wrong thing that was happening before.

**Comment by mvines** [comment]
> Want to land it? This is doing the right thing, vs the wrong thing that was happening before.

Yep, let's get this into v1.2 ASAP for the release this afternoon.  It can't hurt

**Comment by carllin** [comment]
@aeyakovenko @sakridge Some datapoints gathered from pointing the Serum UI at a specific RPC node. Looks like there's no major difference, but agreed the fix is still the correct change to do.

Before this change:
Screen Shot 2020-09-04 at 3 13 20 PM

With this change:
Screen Shot 2020-09-04 at 2 17 27 PM

**Comment by carllin** [comment]
merged in 1.2 here: https://github.com/solana-labs/solana/pull/12012

**Comment by garious** [comment]
@aeyakovenko, instead of duplicating this code and making it so BanksServer doesn't benefit from the optimization, you could have added a trait with a `send_to_tpu(tx)` method and implemented that trait in solana_core. The default impl could send to the local TPU port.

**Comment by aeyakovenko** [comment]
Cargo would have a circular dependency core => bank server => core

**Comment by garious** [comment]
@aeyakovenko, Not if BanksServer had a constructor that accepted that trait

**Pull request merged**